### DISCUSSION TITLE
How to receive keyboard events from Unreal viewport in Airsim

### DISCUSSION THREAD

**Discussion by karnikram** [microsoft/AirSim#3381] [Support Q&A] [answered]
Hi,

Is there a way to receive a key press event in the Unreal viewport, using the Python API? There is currently a `wait_key` function offered, but it only works for key presses made while inside the command console.

I need this because I've configured my Unreal environment to reset upon a key press, for which the focus has to be on the viewport. Upon this reset I want Airsim also to be reset.

**Comment by ahmed-elsaharti** [comment, marked as answer]
Hi @karnikram, have a look at the [keyboard python library](https://github.com/boppreh/keyboard). It lets you capture any keypress regardless of window focus (and much more).

Here's a quick sample showing how you could go about doing that:
```python
import keyboard

def key_press(key):
    if key.name == 'x':
        print("X was pressed")
    if key.name == 'c': 
        print("Reset key pressed")
        client.reset() # or maybe even keyboard.send('backspace')        

keyboard.on_release(key_press)
```
**Note:** in the code above, you just need to include keyboard.on_release(key_press) once in your code, ie: just add it to the beginning of your code, you don't need to loop over it or anything.

**Reply by karnikram** [reply]
Thanks for the suggestion, @ahmed-elsaharti. It seems like the library has issues with [WSL](https://github.com/boppreh/keyboard/issues/394). So I moved our setup to Windows completely and got it to work. Thanks again!
